### PR TITLE
enhancement(types): Enable strictNullChecks and throw errors in unhandled situations

### PR DESCRIPTION
**Pull request by martinjlowm**
This PR enables strictNullChecks and catches undefined behavior. For now, all unhandled situations throw errors with a stack trace, enabling users to supply that in an issue.

If desired I can throw in FIXME's in those places as well.

### CODE REVIEW

**Comment by uittorio** [comment]
Hi @martinjlowm, I appreciate the effort that you put into this. We definitely want to migrate to strictNullChecks = true. 

I honestly would like a better error Message for each scenario. I also think that few scenario can be fixed in a different way. I'll explain a bit more in each line

File: src/extension/extensionHandler.ts [hunk 1]
```diff
@@ -25,6 +25,10 @@ export class ExtensionHandler<TMock> {
       return extensionOrPropertyName(this._mock);
     }
 
+    if (!maybePropertyHandler) {
```
Comment (uittorio) [inline src/extension/extensionHandler.ts:28]: Considering that function overload it's not helping us enough (In my opinion people should not be able to call this method without that parameter unless they disable type checking. What we could do instead is to create a different method in the ExtensionHandler. @Pmyl Tell me If I am wrong :)

File: src/options/options.ts [hunk 2]
```diff
@@ -7,9 +7,9 @@ export interface TsAutoMockOptions {
   cacheBetweenTests: TsAutoMockCacheOptions;
 }
 
-let options: TsAutoMockOptions = null;
+let options: TsAutoMockOptions | undefined;
```
Comment (uittorio) [inline src/options/options.ts:10]: In reality, the options passed to SetTsAutoMockOptions are a partial of TsAutoMockOptions. To avoid spreading undefined  we could do something like this.

```ts
const options: TsAutoMockOptions = defaultOptions;

export function SetTsAutoMockOptions(_options: Partial<TsAutoMockOptions>): void {
  Object.keys(_options).forEach((key: string ) => {
    options[key] = _options[key];
  });
}
export function GetOptionByKey<T extends keyof TsAutoMockOptions>(optionKey: T): TsAutoMockOptions[T] {
  return options[optionKey];
}
```

File: src/transformer/base/base.ts [hunk 3]
```diff
@@ -42,13 +42,17 @@ function visitNode(node: ts.Node, visitor: Visitor, customFunctions: CustomFunct
...
 
-  if (!isFunctionFromThisLibrary(signature, customFunctions)) {
+  if (!signature || !isFunctionFromThisLibrary(signature, customFunctions)) {
     return node;
   }
 
-  const nodeToMock: ts.TypeNode = node.typeArguments[0];
+  const nodeToMock: ts.TypeNode | undefined = node.typeArguments?.[0];
+
+  if (!nodeToMock) {
```
Comment (uittorio) [inline src/transformer/base/base.ts:53]: Again. this would mean calling createMock without a generic. This means disabling the typecheck. If consumers are willing to use it we should throw a more specific error :) Something like. Hei, it looks like you are using {whatever} without a generic? check the documentation at this link

File: src/transformer/descriptor/helper/helper.ts [hunk 4]
```diff
@@ -16,7 +16,11 @@ export namespace TypescriptHelper {
 
   export function GetDeclarationFromNode(node: ts.Node): ts.Declaration {
     const typeChecker: ts.TypeChecker = TypeChecker();
-    const symbol: ts.Symbol = typeChecker.getSymbolAtLocation(node);
+    const symbol: ts.Symbol | undefined = typeChecker.getSymbolAtLocation(node);
```
Comment (uittorio) [inline src/transformer/descriptor/helper/helper.ts:19]: So, this could happen, true. I am not sure when. We didn't have any issues with this yet. (Cannot read name of undefined). I would either investigate when this could happen or I would put some more useful error message.

File: src/create-mock.ts [hunk 5]
```diff
@@ -1,3 +1,3 @@
 import { DeepPartial } from './partial/deepPartial';
 
-export declare function createMock<T extends object>(values?: DeepPartial<T>): T;
+export declare function createMock<T extends object | undefined>(values?: DeepPartial<T>): T;
```
Comment (uittorio) [inline src/create-mock.ts:3]: If you check my comment below I wouldn't set this as undefined unless there is a specific reason

**Comment by uittorio** [comment]
I left a few comments, I didn't check all of them. Ideally, we want to find the cause of the undefined or null first. I'm okay to start putting this error in place but we need a better error message. What do you think?

**Comment by martinjlowm** [comment]
@uittorio thanks for your comments. I will probably find time tonight (CEST time) to iterate on them and hopefully have some adjustments ready by tomorrow.

It's easier with some feedback to determine what the intended outcome is, especially with that in mind that I started exploring the code base two days ago. So, thank you for that! I agree with your proposal regarding better error messages, I just thought an error is better than unexpected behavior and making the type checker happy.

Also, Happy Easter! :)

File: src/create-mock.ts (hunk 5 above)
Comment (martinjlowm) [reply src/create-mock.ts:3]: This was solely intended for the type query test covering `createMock<typeof undefined>`, there, `object` isn't enough. The meaning of it, I'm not sure 😛

**Comment by uittorio** [comment]
@martinjlowm HI :) Happy Easter

Oh don't' worry, After 2 days of exploration you have done a great job.

I Agree with you, an Error is better than unexpected behavior but previously if there was an error the stack would be more clear than the new Error, that's why we have to be a bit more specific to be at least clear as before.

File: src/extension/extensionHandler.ts (hunk 1 above)
Comment (martinjlowm) [reply src/extension/extensionHandler.ts:28]: I presume making the property handler (`maybePropertyHandler`) required is the desired behavior in this case?

File: src/create-mock.ts (hunk 5 above)
Comment (uittorio) [reply src/create-mock.ts:3]: I would prefer to remove the eslint in the test. The usual behavior is always with an object.

File: src/options/options.ts [hunk 6]
```diff
@@ -7,16 +7,15 @@ export interface TsAutoMockOptions {
   cacheBetweenTests: TsAutoMockCacheOptions;
 }
 
-let options: TsAutoMockOptions = null;
+let tsAutoMockOptions: TsAutoMockOptions = defaultOptions;
 
-export function SetTsAutoMockOptions(_options: TsAutoMockOptions): void {
-  options = _options;
+export function SetTsAutoMockOptions(options?: TsAutoMockOptions): void {
```
Comment (uittorio) [inline src/options/options.ts:12]: You could remove the question mark and do an if statement in the transformer. if(options) { SetTsAUtoMockOptions(options)}

File: src/extension/extensionHandler.ts (hunk 1 above)
Comment (uittorio) [reply src/extension/extensionHandler.ts:28]: it should be required but because is overloading it's not. I am happy with the error. Something similar to the createMock one. "It looks like you are trying to get an extension for {propertyName} without specifying the handler.

File: src/transformer/descriptor/mock/mockProperties.ts [hunk 7]
```diff
@@ -19,21 +19,21 @@ export function GetMockPropertiesFromSymbol(propertiesSymbol: ts.Symbol[], signa
 
 export function GetMockPropertiesFromDeclarations(list: ReadonlyArray<PropertyLike>, signatures: ReadonlyArray<SignatureLike>, scope: Scope): ts.CallExpression {
   const propertiesFilter: PropertyLike[] = list.filter((member: PropertyLike) => {
-    const hasModifiers: boolean = !!member.modifiers;
+    const { modifiers = ts.createNodeArray([]) }: PropertyLike = member;
```
Comment (uittorio) [inline src/transformer/descriptor/mock/mockProperties.ts:22]: these changes are not necessary. Could we revert these changes in this PR?

File: src/transformer/mockDefiner/mockDefiner.ts [hunk 8]
```diff
@@ -213,8 +232,9 @@ export class MockDefiner {
   private _getMockFactoryIdForIntersections(declarations: ts.Declaration[], intersectionTypeNode: ts.IntersectionTypeNode): string {
     const thisFileName: string = this._fileName;
 
-    if (this._factoryIntersectionCache.has(declarations)) {
-      return this._factoryIntersectionCache.get(declarations);
+    const factoryIntersection: string | undefined = this._factoryIntersectionCache.get(declarations);
```
Comment (uittorio) [inline src/transformer/mockDefiner/mockDefiner.ts:235]: I preferred the old version. Does it complain typescript? Can we cast it considering that if has works the get should work?

File: src/transformer/mockDefiner/mockDefiner.ts [hunk 9]
```diff
@@ -250,10 +270,12 @@ export class MockDefiner {
     if (this._factoryRegistrationsPerFile[sourceFile.fileName]) {
       return this._factoryRegistrationsPerFile[sourceFile.fileName]
         .map((reg: { key: ts.Declaration; factory: ts.Expression }) => {
-          const key: string = this._factoryCache.get(reg.key);
+          const key: string | undefined = this._factoryCache.get(reg.key);
```
Comment (uittorio) [inline src/transformer/mockDefiner/mockDefiner.ts:273]: This cannot fail. We should improve the types

File: src/transformer/mockDefiner/mockDefiner.ts [hunk 10]
```diff
@@ -263,10 +285,12 @@ export class MockDefiner {
     if (this._factoryIntersectionsRegistrationsPerFile[sourceFile.fileName]) {
       return this._factoryIntersectionsRegistrationsPerFile[sourceFile.fileName]
         .map((reg: { keys: ts.Declaration[]; factory: ts.Expression }) => {
-          const key: string = this._factoryIntersectionCache.get(reg.keys);
+          const key: string | undefined = this._factoryIntersectionCache.get(reg.keys);
```
Comment (uittorio) [inline src/transformer/mockDefiner/mockDefiner.ts:288]: Same as above

File: src/transformer/mockDefiner/mockDefiner.ts [hunk 11]
```diff
@@ -276,10 +300,12 @@ export class MockDefiner {
     if (this._registerMockFactoryRegistrationsPerFile[sourceFile.fileName]) {
       return this._registerMockFactoryRegistrationsPerFile[sourceFile.fileName]
         .map((reg: { key: ts.Declaration; factory: ts.Expression }) => {
-          const key: string = this._registerMockFactoryCache.get(reg.key);
+          const key: string | undefined = this._registerMockFactoryCache.get(reg.key);
```
Comment (uittorio) [inline src/transformer/mockDefiner/mockDefiner.ts:303]: same as above

File: src/transformer/transformer.ts [hunk 12]
```diff
@@ -26,31 +26,33 @@ const transformer:
 export { transformer };
 
 function visitNode(node: ts.CallExpression, declaration: ts.FunctionDeclaration): ts.Node {
-  const nodeToMock: ts.TypeNode = node.typeArguments[0];
+  const nodeToMock: ts.TypeNode | undefined = node.typeArguments?.[0];
```
Comment (uittorio) [inline src/transformer/transformer.ts:29]: Considering that this visitNode already passed the check of typeArguments, I would create an interface that extends ts.CallExpression that always as typeArguments. Is that possible?

File: test/transformer/create-mock-values.test.ts [hunk 13]
```diff
@@ -6,9 +6,9 @@ describe('create-mock-values', () => {
   }
 
   interface Interface {
-    property: string;
-    method(): void;
-    subInterface: SubInterface;
+    property: string | null;
```
Comment (uittorio) [inline test/transformer/create-mock-values.test.ts:9]: Why do we need null?

File: src/transformer/mockDefiner/mockDefiner.ts (hunk 8 above)
Comment (martinjlowm) [reply src/transformer/mockDefiner/mockDefiner.ts:235]: I changed it because `has` doesn't act like a type guard and the `get` would result in `string | undefined` in either case. One thing I missed though, is that the change doesn't consider empty strings - however, I don't think that is a valid value anyway. `has` does check that implicitly.

File: src/options/options.ts (hunk 6 above)
Comment (martinjlowm) [reply src/options/options.ts:12]: That makes sense and I agree, it should be the user's responsibility to pass in a truthy value.

File: src/transformer/descriptor/mock/mockProperties.ts (hunk 7 above)
Comment (martinjlowm) [reply src/transformer/descriptor/mock/mockProperties.ts:22]: Sure. I guess doing an unnecessary `ts.createNodeArray([])` call was a bit of a stretch. The goal is to type guard modifiers or tell the type checker that we know what we're doing and add an exclamation mark on `modifiers` in the return statement.

File: src/transformer/transformer.ts (hunk 12 above)
Comment (martinjlowm) [reply src/transformer/transformer.ts:29]: I guess we can adjust the signature to something like `ts.CallExpression & { typeArguments: NonNullable<ts.CallExpression['typeArguments']> }`

File: test/transformer/create-mock-values.test.ts (hunk 13 above)
Comment (martinjlowm) [reply test/transformer/create-mock-values.test.ts:9]: `property` (and the other properties) of `Interface` is assigned null in the `should create the mock merging the null values provided` test and they are asserted as such too.

File: src/transformer/transformer.ts (hunk 12 above)
Comment (uittorio) [reply src/transformer/transformer.ts:29]: I think we could use (node: ts.CallExpression & { typeArguments: NodeArray<TypeNode> })

File: src/transformer/descriptor/mock/mockProperties.ts (hunk 7 above)
Comment (martinjlowm) [reply src/transformer/descriptor/mock/mockProperties.ts:22]: On that note, the ESLint config disallows the use of `!` so we would have to cast it instead. I think type guarding is cleaner though and more explicit. `isTruthyValue` here can be used in more places than this, however, the requirement of variable typing would be a bit tedious (since `hasModifiers` is now a function).

```
const isTruthyValue = <T>(arg: T): arg is NonNullable<T> => {
  return !!c;
};

const hasModifiers = isTruthyValue;

if (!hasModifiers(member.modifiers)) {
  return true;
}

return member.modifiers.filter((modifier: ts.Modifier) => modifier.kind === ts.SyntaxKind.PrivateKeyword).length === 0;
```

Comment (uittorio) [reply src/transformer/descriptor/mock/mockProperties.ts:22]: I see, let's simply assign modifiers to a variable. 

```ts
const modifiers: ts.NodeArray<ts.Modifier> | undefined = member.modifiers;

    if (IsTypescriptType(member)) { // This is a current workaround to safe fail extends of TypescriptLibs
      return false;
    }

    if (!modifiers) {
      return true;
    }

    return modifiers.filter((modifier: ts.Modifier) => modifier.kind === ts.SyntaxKind.PrivateKeyword).length === 0;
```

File: src/transformer/mockDefiner/mockDefiner.ts (hunk 8 above)
Comment (martinjlowm) [reply src/transformer/mockDefiner/mockDefiner.ts:235]: Perhaps, we should allow the ESLint `@typecript-eslint/no-non-null-assertion`-rule. That would allow us to do: `this._factoryIntersectionCache.get(declarations)!`.

File: src/transformer/mockDefiner/mockDefiner.ts (hunk 9 above)
Comment (martinjlowm) [reply src/transformer/mockDefiner/mockDefiner.ts:273]: Disabling `@typecript-eslint/no-non-null-assertion` would allow us to replace this (and the other places) with the following if we can guarantee that the key will be available at all times.

```
const key: string = this._factoryCache.get(reg.key)!;
```
